src/libraries/System.Linq/tests/ElementAtTests.cs

```diff
@@ -72,12 +104,426 @@ public void NullableArray_ValidIndex_ReturnsCorrectObject()
 
             Assert.Null(source.ElementAt(2));
             Assert.Equal(-5, source.ElementAt(3));
+
+            Assert.Null(source.ElementAt(new Index(2)));
+            Assert.Equal(-5, source.ElementAt(new Index(3)));
+
+            Assert.Null(source.ElementAt(^3));
+            Assert.Equal(-5, source.ElementAt(^2));
         }
 
         [Fact]
         public void NullSource_ThrowsArgumentNullException()
         {
             AssertExtensions.Throws<ArgumentNullException>("source", () => ((IEnumerable<int>)null).ElementAt(2));
+            AssertExtensions.Throws<ArgumentNullException>("source", () => ((IEnumerable<int>)null).ElementAt(new Index(2)));
+            AssertExtensions.Throws<ArgumentNullException>("source", () => ((IEnumerable<int>)null).ElementAt(^2));
+        }
+
+        [Fact]
+        public void MutableSource()
+        {
+            var source = new List<int>() { 0, 1, 2, 3, 4 };
+            Assert.Equal(2, source.ElementAt(2));
+            Assert.Equal(2, source.ElementAt(new Index(2)));
+            Assert.Equal(2, source.ElementAt(^3));
+
+            source.InsertRange(3, new[] { -1, -2 });
+            source.RemoveAt(0);
+            Assert.Equal(-1, source.ElementAt(2));
+            Assert.Equal(-1, source.ElementAt(new Index(2)));
+            Assert.Equal(-1, source.ElementAt(^4));
+        }
+
+        [Fact]
+        public void MutableSourceNotList()
+        {
+            var source = new List<int>() { 0, 1, 2, 3, 4 };
+            var query1 = Repeat(_ => ForceNotCollection(source).Select(i => i), 3);
+            Assert.Equal(2, query1[0].ElementAt(2));
+            Assert.Equal(2, query1[1].ElementAt(new Index(2)));
+            Assert.Equal(2, query1[2].ElementAt(^3));
+
+            var query2 = Repeat(_ => ForceNotCollection(source).Select(i => i), 3);
+            source.InsertRange(3, new[] { -1, -2 });
+            source.RemoveAt(0);
+            Assert.Equal(-1, query2[0].ElementAt(2));
+            Assert.Equal(-1, query2[1].ElementAt(new Index(2)));
+            Assert.Equal(-1, query2[2].ElementAt(^4));
+        }
+
+        private class ArrayWrapper<T> : IList<T>
```

Comment:
I'm not sure I understand why this wrapper is needed? Arrays already implement `IList<T>` and that is how `ElementAt` consumes them.

Reply:
For an array: 
- `array[-1]` throws `IndexOutOfRangeException`
- `((IList<T>)array))[-1]` throws `ArgumentOutOfRangeException`.

For ElementAt:
- When array is passed to ElementAt, it calls array's `IList<T>` indexer, and throws ArgumentOutOfRangeException.
- When ArrayWrapper is passed to ElementAt, it calls ArrayWrapper indexer, which then calls the real array indexer, and throws IndexOutOfRangeException. 
  - This tests the case when the user implements a custom collection type, with an indexer throwing exception other than ArgumentOutOfRangeException. 

Typically, when user implements a CustomCollection type, they could usually implement it based on array. And CustomCollection's indexer usually calls array[index]. So:
- array.ElementAt(indexOutOfBounds) throws ArgumentOutOfException
- arrayBasedCustomCollection.ElementAt(indexOutOfBounds) usually throws IndexOutOfRangeException. Example: System.Collections.ImmutableArray, etc.

I will add comments to ArrayWrapper.

Reply:
Would it make sense to write a mock implementation that simply throws a specific exception on the indexer? I think it would be more self-documenting.